sdk/java/src/main/java/net/consensys/gpact/messaging/common/attestorrelayer/AttestorRelayer.java

```diff
@@ -0,0 +1,134 @@
+    String crosschainControlAddr;
+    String observerUri;
+    String contractType;
+    String bcWsUri;
+
+    public Source(
+        BlockchainId bcId,
+        String crosschainControlAddr,
+        String contractType,
+        String observerUri,
+        String bcWsUri) {
+      this.bcId = bcId;
+      this.crosschainControlAddr = crosschainControlAddr;
+      this.contractType = contractType;
+      this.observerUri = observerUri;
+      this.bcWsUri = bcWsUri;
+    }
+
+    public String getSourceId() {
+      return this.bcId.toString() + this.crosschainControlAddr;
+    }
+  }
+
+  Map<String, Source> sources = new HashMap<>();
+
+  public AttestorRelayer(String relayerUri, byte[] signingKey) {
+    this.relayerUri = relayerUri;
+    this.signingKey = signingKey;
+  }
+
+  public AttestorRelayer(String msgStoreAddresses) {
+    this.msgStoreAddr = msgStoreAddresses;
+  }
+
+  public void addNewSource(
+      BlockchainId bcId,
+      String crosschainControlAddr,
+      String contractType,
+      String observerUri,
+      String bcWsUri)
+      throws CrosschainProtocolStackException {
+    addNewSource(new Source(bcId, crosschainControlAddr, contractType, observerUri, bcWsUri));
+  }
+
+  public void addNewSource(Source source) throws CrosschainProtocolStackException {
+    this.sources.put(source.getSourceId(), source);
+
+    AttestorRelayerWebApi.setupObserver(
+        source.observerUri,
+        source.bcId,
+        source.bcWsUri,
+        source.contractType,
+        source.crosschainControlAddr);
+    AttestorRelayerWebApi.setupRelayer(
+        this.relayerUri, source.bcId, source.crosschainControlAddr, this.signingKey);
+  }
+
+  public void addMessageStore(
+      String msgDispatcherUrl, String msgStoreAddrFromDispatcher, String msgStoreAddrFromUser)
+      throws CrosschainProtocolStackException {
+    AttestorRelayerWebApi.setupDispatcher(msgDispatcherUrl, msgStoreAddrFromDispatcher);
+    this.msgStoreAddr = msgStoreAddrFromUser;
+  }
+
+  public String fetchedSignedEvent(
+      BlockchainId bcId,
+      TransactionReceipt txr,
+      String contractAddress,
+      byte[] eventFunctionSignature)
+      throws CrosschainProtocolStackException {
+    String eventId = calculateEventID(bcId, txr, contractAddress, eventFunctionSignature);
+
+    return AttestorRelayerWebApi.fetchSignedEvent("http://" + this.msgStoreAddr, eventId);
+  }
+
+  public static String calculateEventID(
+      BlockchainId bcId,
+      TransactionReceipt txr,
+      String contractAddress,
+      byte[] eventFunctionSignature) {
+    String chainId = bcId.toDecimalString();
+
+    List<Log> logs = txr.getLogs();
+    for (Log log : logs) {
+      if (log.getAddress().equalsIgnoreCase(contractAddress)
+          && log.getTopics()
+              .get(0)
+              .equalsIgnoreCase(FormatConversion.byteArrayToString(eventFunctionSignature))) {
+
+        String eventAddr = "0x" + log.getAddress().toLowerCase().substring(2);
```

Comment:
nit: is the substring pruning out "0x" here? if so, is that redundant given we are concatenating "0x" again?

Reply:
This needs to be removed. This was to handle the old code that took the string to upper case, but then needed to "x" to be lower case. At that point, I didn't realise the old Relayer code was using EIP 55 mixed case.

Reply:
Fixed